File: articles/nonlinearity-feedback-biological-regulation/python/logistic_regulation.py
"""
Logistic density-dependent regulation scenarios.

Run:
    python python/logistic_regulation.py
"""

from pathlib import Path

import pandas as pd

from nonlinear_feedback_core import simulate_logistic


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "logistic_regulation_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        trajectory = simulate_logistic(row["N0"], row["r"], row["K"], row["dt"], row["t_end"])

        rows.append(
            {
                "scenario": row["scenario"],
                "final_population": trajectory["population"].iloc[-1],
                "fraction_of_capacity": trajectory["population"].iloc[-1] / row["K"],
                "max_population": trajectory["population"].max(),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/nonlinearity-feedback-biological-regulation/r/logistic_regulation.R
# Logistic density-dependent regulation workflow.

simulate_logistic <- function(N0, r, K, dt = 0.05, t_end = 40) {
  time <- seq(0, t_end, by = dt)
  N <- numeric(length(time))
  N[1] <- N0

  for (i in 2:length(time)) {
    dN <- r * N[i - 1] * (1 - N[i - 1] / K)
    N[i] <- max(N[i - 1] + dN * dt, 0)
  }

  data.frame(time = time, population = N)
}

trajectory <- simulate_logistic(100, 0.30, 2000)

summary_df <- data.frame(
  initial_population = trajectory$population[1],
  final_population = tail(trajectory$population, 1),
  fraction_of_capacity = tail(trajectory$population, 1) / 2000
)

print(round(summary_df, 5))
